### PR TITLE
Framework for mesolve and odeint

### PR DESCRIPTION
**Pull request by gautierronan**
In this PR, I make an attempt at the overall framework for mesolve (and sesolve, smesolve) and odeint, following the discussion we had the other day.

I introduce a class named `QSolver` which implements a single ODE iteration of any given solver. For now, it is written for first and second-order Rouchon methods.  It can later also be derived into subclasses for other solvers, both fixed and adaptive time steps. It can also easily be derived with the 1.5 order Rouchon scheme of https://github.com/PierreGuilmin/torchqdynamics/pull/5. All Qsolvers will eventually support the `forward_adjoint` method for adjoint-based autodiff. 

Inside `mesolve`, an instance of `QSolver` is created using the solver options dataclass of https://github.com/PierreGuilmin/torchqdynamics/pull/10, and the ODE integrator `odeint` is called. In the end, `odeint` is meant to be common to all main functions of the library, such as `sesolve` and `smesolve`.

I intentionnally made the Hamiltonian `H` time-dependent for now, and the `jump_ops` time-independent, as I believe this should be the most common scenario. Also, `H` is a callable for now, but meant to also accept a `torch.Tensor` or a `List[torch.Tensor]` with time-dependent coefficients in the end. This should be done in a separate PR.

Anyway, I'll let you read this. I tried to be as basic as possible so we can all agree on the framework and then build on it. I found this was a good way to meet all of our demands. I will personnally be able to build on this for the adjoint-based odeint.

### CODE REVIEW

File: torchqdynamics/mesolve.py
Comment (abocquet) [inline torchqdynamics/mesolve.py:103]: Using strings for enumeration is a pattern I'm ill at ease with because we don't know what are the accepted values. Here I'm guessing `stepclass` could be `fixed` or "fixed but variable in time"?

**Review by abocquet — COMMENTED** [review]
Only small nits for now. I have to think a little more about the architecture, I think we had different things in mind

File: torchqdynamics/mesolve.py
Comment (gautierronan) [reply torchqdynamics/mesolve.py:13]: Yes I tried to stay as simple as possible for this. But it for sure should be added later.

File: torchqdynamics/mesolve.py
Comment (gautierronan) [reply torchqdynamics/mesolve.py:103]: I was thinking that `stepclass` should be either `fixed` or `adaptive` here. It's just an information so that we know whether we should dispatch to the `fixed_odeint` or `adaptive_odeint` later on. In your solver options PR you have the two separate classes for fixed and adaptive, so we could use just the parent class name instead of a string class argument?

For time-dependence, I think it should be implemented directly in the QSolver. Something like `MERouchon1` and `MERouchon1_TD` maybe? Or it could be the same class, but just pre-compute the Kraus operators in the `Qsolver.__init__`?

Comment (abocquet) [reply torchqdynamics/mesolve.py:102]: +1 on the separate classes if needed, but I thought the Rouchon method could only support fixed time steps (or time dependent but fixed in advanced).

If all fixed timestep method support both a simple dt and a time depedent timestep then we could simply say that the `step` argument accepts a float or a list of floats?

Comment (gautierronan) [reply torchqdynamics/mesolve.py:102]: No that's not what I meant the `stepclass` argument for. I meant it to be used here: https://github.com/PierreGuilmin/torchqdynamics/blob/f00ac9e980515c434891cc2c394e81eecbbd2198/torchqdynamics/odeint.py#L31-L34
just to dispatch to the related fixed or adaptive odeint solver. It's unrelated to time-dependence or varying dt.

Comment (abocquet) [reply torchqdynamics/mesolve.py:102]: Yes, agreed. Are there methods that can be both used with a fixed timestep or an adaptative one?
For instance I think that Rouchon method is fixed only and DOPRI6 is adaptative only?

Comment (gautierronan) [reply torchqdynamics/mesolve.py:102]: No I think it's always one or the other.

Comment (abocquet) [reply torchqdynamics/mesolve.py:102]: Then can't be `stepclass` inferred from the schema?

Comment (gautierronan) [reply torchqdynamics/mesolve.py:102]: Yes but I didn't know how to, and I also wanted a simple implementation for this PR. But it should be implemented properly from your PR instead.

Comment (abocquet) [reply torchqdynamics/mesolve.py:102]: Ok good to know, I just wanted to make sure I didn't miss something.
Let's do this after we merge https://github.com/PierreGuilmin/torchqdynamics/pull/10

**Review by abocquet — APPROVED** [review]

File: torchqdynamics/mesolve.py
Comment (gautierronan) [inline torchqdynamics/mesolve.py:79]: (NB: Bug here needs to be fixed. Line 79 should be in line 78 without the adjoint)

File: torchqdynamics/odeint.py
Comment (gautierronan) [inline torchqdynamics/odeint.py:18]: A good solution here might be to simply use
```python
if sensitivity is None:
    with torch.no_grad():
        return odeint_main(qsolver, y0, tsave)
```
This way torch does not store gradients. It still might be slower than an in-place solver, but it should give a performance boost.

**Review by PierreGuilmin — DISMISSED** [review]
Thanks a lot for this great PR @gautierronan! It looks like a very good starting point for the rest 🥳 I did mostly minor changes, I advise you to check the commits one by one. I am happy to discuss anything that looks weird.

Various comments:
- I merged the `main` branch after having accepted #10. The recommended workflow is to directly base your work on top of #10 instead of implementing dummy code to fake it 😉 For that you just need to checkout the branch `abocquet/solvers-options` of #10, then create a new branch from there, and then when opening a PR on GitHub ask to merge into the branch `abocquet/solvers-options` instead of `main`. Then when we merge #10 GitHub will automatically change the target branch of the PR to the `main` branch.
- Commit 758f8e752cf095d2d83a6caf8d77c5e3c43993e8 (hover to see commit title) fixes various style issues with the comments (to match common Python style of the main libraries), namely
  - Remove file comments (the docstrings and file names/organisations **must be** sufficient).
  - Inline comments start with a lowercase, except if it's a long explanation needing multiple sentences spanning multiple lines.
  - Add final dots to docstring comments.
  - Add inline code formatting when needed.
- 1ab48f55373bf2b21aa8dc2f0de086d6b2e98cbc because the second order is bugged as you pointed out, I'd rather have this in a separate clean PR.

**Comment by PierreGuilmin** [comment]
I'm good to merge this PR if you agree @gautierronan and @abocquet, and then we can port the nice extras from Adrien's Rouchon solver implementation (#5).

**Review by PierreGuilmin — APPROVED** [review]
Let's go 🥳

**Pull request merged**
